lib/shape.ml

```diff
@@ -15,147 +13,64 @@ type shape =
-  (int_of_float nx, int_of_float ny)
+(* point + scalar arithmetic *)
+let ( -! ) { x = x1; y = y1 } scalar = { x = x1 -. scalar; y = y1 -. scalar }
+let splat n = { x = n; y = n }
 
-let denormalize point =
-  { x = point.x + canvas_mid.x; y = point.y + canvas_mid.y }
+let bi_to_uni { x; y } =
+  let cx, cy = Context.resolution () in
+  let nx = (x *. 0.5) +. (cx *. 0.5) in
+  let ny = (y *. 0.5) +. (cy *. 0.5) in
+  (nx, ny)
 
-let render_polygon polygon' =
-  let points =
-    Array.of_list
-      (List.map
-         (fun point ->
-           let { x; y } = denormalize point in
-           (x, y))
-         polygon')
-  in
-  draw_poly points
+let denormalize point =
+  let x, y = Context.resolution () in
+  let canvas_mid = { x; y } /~ splat 2. in
+  { x = point.x +. canvas_mid.x; y = point.y +. canvas_mid.y }
 
-let rec render_shape s =
-  match s with
-  | Circle circle ->
-      draw_circle (denormalize circle.c).x (denormalize circle.c).y
-        circle.radius
-  | Ellipse ellipse ->
-      let c = denormalize ellipse.c in
-      draw_ellipse c.x c.y ellipse.rx ellipse.ry
-  | Line line ->
-      let a = denormalize line.a in
-      let b = denormalize line.b in
-      draw_line a.x a.y b.x b.y
-  | Polygon polygon' -> render_polygon polygon'
-  | Complex complex -> List.iter render_shape complex
+(* Scales points from 0-image size to 0-1 on both axes *)
+let scale_point size point =
+  let { x; y } = denormalize point in
+  let x, y = (x /. fst size, y /. snd size) in
+  (x, y)
 
 let point x y = { x; y }
 
 let circle ?point r =
   match point with
   | Some point -> Circle { c = point; radius = r }
-  | _ -> Circle { c = { x = 0; y = 0 }; radius = r }
+  | _ -> Circle { c = { x = 0.; y = 0. }; radius = r }
 
-let rectangle ?point length width =
-  let x, y = match point with Some { x; y } -> (x, y) | None -> (0, 0) in
+let make_rectangle c width height =
+  let width, height = (width *. 2., height *. 2.) in
+  let { x; y } = c -! ((width +. height) /. 4.) in
   Polygon
     [
       { x; y };
-      { x; y = y + length };
-      { x = x + width; y = y + length };
-      { x = x + width; y };
+      { x; y = y +. height };
+      { x = x +. width; y = y +. height };
+      { x = x +. width; y };
     ]
 
+let rectangle ?point length width =
+  make_rectangle
+    (match point with Some c -> c | None -> { x = 0.; y = 0. })
+    length width
+
 let ellipse ?point rx ry =
   match point with
-  | Some point -> Ellipse { c = point; rx; ry }
-  | _ -> Ellipse { c = { x = 0; y = 0 }; rx; ry }
+  | Some c -> Ellipse { c; rx; ry }
+  | None -> Ellipse { c = { x = 0.; y = 0. }; rx; ry }
 
-let line ?point_a point_b =
-  match point_a with
-  | Some point_a -> Line { a = point_a; b = point_b }
-  | _ -> Line { a = { x = 0; y = 0 }; b = point_b }
+let line ?point point_b =
+  match point with
+  | Some p -> Line { a = p; b = point_b }
+  | _ -> Line { a = { x = 0.; y = 0. }; b = point_b }
```

Comment:
My intention here was to keep the name of the optional point argument the same across all shape constructor functions, so that hopefully users don't have to try remember the name of the optional arg for each function. Do you want to do 'point' and 'b' maybe? or do you think 'a' and 'b' would be preferable?